Add a new chapter in the pong tutorial. (#1061)
amethyst/amethyst: book/src/pong-tutorial/pong-tutorial-04.md

# Making a ball move and bounce

In the previous chapter, we learned how to capture user input

> to make things move on the screen by creating a so-called System

Reviewer: ```suggestion
to make things move on the screen by creating a so-called `System`
```

ourselves.
This chapter will reuse all the knowledge we acquired through the
previous chapters to add a new object to our game: a ball that moves
and bounces around!

First, let's define some other useful constants for this chapter:
```rust,no_run,noplaypen
pub const BALL_VELOCITY_X: f32 = 75.0;
pub const BALL_VELOCITY_Y: f32 = 50.0;
pub const BALL_RADIUS: f32 = 2.0;
```

> 

Reviewer: Maybe point out this could be placed in a config file for when you want to edit these values a lot? Or probably the pong example should be changed to using that?

Author: I don't know if it is sensible to add this comment here as it is the way it's done in previous chapters. But I may update those as well? Also, what about making a chapter for this kind of "refactors" later?

Reviewer: :+1: I think a simple pointer here would be enough.

Reviewer: something like "This could also be done by .., but we're keeping it simple here."

# const BALL_RADIUS: f32 = 2.0;
# const BALL_VELOCITY_X: f32 = 75.0;
# const BALL_VELOCITY_Y: f32 = 50.0;
# const ARENA_HEIGHT: f32 = 100.0;
# const ARENA_WIDTH: f32 = 100.0;
/// Initialises one ball in the middle-ish of the arena.
fn initialise_ball(world: &mut World, sprite_sheet_handle: SpriteSheetHandle) {
    // Create the translation.
    let mut local_transform = Transform::default();
    local_transform.translation = Vector3::new(ARENA_WIDTH / 2.0, ARENA_HEIGHT / 2.0, 0.0);

    // Assign the sprite for the ball
    let sprite_render = SpriteRender {
        sprite_sheet: sprite_sheet_handle,
        sprite_number: 1, // ball is the second sprite on the sprite_sheet
        flip_horizontal: true,
        flip_vertical: false,
    };

    world
        .create_entity()
        .with(sprite_render)
        .with(Ball {
            radius: BALL_RADIUS,
            velocity: [BALL_VELOCITY_X, BALL_VELOCITY_Y],
        })
        .with(local_transform)
        .build();
}
```

In [a previous chapter][pong_02_drawing] we saw how to load a sprite sheet
and get things drawn on the screen. As you may remember sprite sheet information

> are stored in `pong_spritesheet.ron` and the ball sprite was the

Reviewer: ```suggestion
is stored in `pong_spritesheet.ron` and the ball sprite was the
```

> second one whose index is `1`.

Reviewer: Sentence seems a little unstructured to me, maybe missing a comma?

# use amethyst::ecs::prelude::{Component, DenseVecStorage};
# pub struct Ball {
#    pub velocity: [f32; 2],
#    pub radius: f32,
# }
# impl Component for Ball {
#    type Storage = DenseVecStorage<Self>;
# }

use amethyst::{
    core::timing::Time,
    core::transform::Transform,
    ecs::prelude::{Join, Read, ReadStorage, System, WriteStorage},
};

pub struct MoveBallsSystem;

impl<'s> System<'s> for MoveBallsSystem {
    type SystemData = (
        ReadStorage<'s, Ball>,
        WriteStorage<'s, Transform>,
        Read<'s, Time>,
    );

    fn run(&mut self, (balls, mut locals, time): Self::SystemData) {
        // Move every ball according to its speed, and the time passed.
        for (ball, local) in (&balls, &mut locals).join() {
            local.translation[0] += ball.velocity[0] * time.delta_seconds();
            local.translation[1] += ball.velocity[1] * time.delta_seconds();
        }
    }
}
```

Note: You will need to add a `use` statement to bring in `Ball` from `pong.rs`.

This system is responsible for moving all balls according to their speed and
the elapsed time.
As you can see, to gain access to time passed since the last frame, you need

> to use the [`amethyst::core::timing::Time`][doc_time], a commonly used

Reviewer: ```suggestion
to use [`amethyst::core::timing::Time`][doc_time], a commonly used
```

resource. It has a method called `delta_seconds` that does exactly what we want.

> Then, we'll implement the `BounceSystem` in `systems/bounce.rs`:

Reviewer: ```suggestion
Next, let's implement the `BounceSystem` in `systems/bounce.rs`:
```

        // from occurring.
        for (ball, transform) in (&mut balls, &transforms).join() {
            let ball_x = transform.translation[0];
            let ball_y = transform.translation[1];

            // Bounce at the top or the bottom of the arena.
            if ball_y <= ball.radius && ball.velocity[1] < 0.0 {
                ball.velocity[1] = -ball.velocity[1];
            } else if ball_y >= ARENA_HEIGHT - ball.radius && ball.velocity[1] > 0.0 {
                ball.velocity[1] = -ball.velocity[1];
            }

            // Bounce at the paddles.
            for (paddle, paddle_transform) in (&paddles, &transforms).join() {
                let paddle_x = paddle_transform.translation[0] - paddle.width * 0.5;
                let paddle_y = paddle_transform.translation[1] - paddle.height * 0.5;

                // To determine whether the ball has collided with a paddle, we create a larger
                // rectangle around the current one, by subtracting the ball radius from the
                // lowest coordinates, and adding the ball radius to the highest ones. The ball
                // is then within the paddle if its centre is within the larger wrapper

>                 // rectangle.

Reviewer 2: If you've got time, maybe a picture could help here?
It's okay if you don't, I just want to make sure it's perfectly clear what we are doing.

                if point_in_rect(
                    ball_x,
                    ball_y,
                    paddle_x - ball.radius,
                    paddle_y - ball.radius,
                    paddle_x + paddle.width + ball.radius,
                    paddle_y + paddle.height + ball.radius,
                ) {
                    if paddle.side == Side::Left && ball.velocity[0] < 0.0 {
                        ball.velocity[0] = -ball.velocity[0];
                    } else if paddle.side == Side::Right && ball.velocity[0] > 0.0 {
                        ball.velocity[0] = -ball.velocity[0];
                    }
                }
            }
        }
    }
}

// A point is in a box when its coordinates are smaller or equal than the top
// right, but larger or equal than the bottom left.
fn point_in_rect(x: f32, y: f32, left: f32, bottom: f32, right: f32, top: f32) -> bool {
    x >= left && x <= right && y >= bottom && y <= top
}
```

This system is responsible for detecting collisions between balls and
paddles, as well as balls and the top and bottom edges of the arena.
If a collision is detected, the ball bounces off. This is done

> by modyfing the velocity of the `Ball` component.

Reviewer 2: Could you explain precisely that it is about negating the velocity?
And maybe move that paragraph before the code so that the reader has their mind ready.

# }
# pub struct MoveBallsSystem;
# impl<'a> amethyst::ecs::System<'a> for MoveBallsSystem {
# type SystemData = ();
# fn run(&mut self, _: Self::SystemData) { }
# }
# pub struct BounceSystem;
# impl<'a> amethyst::ecs::System<'a> for BounceSystem {
# type SystemData = ();
# fn run(&mut self, _: Self::SystemData) { }
# }
# }
# let input_bundle = amethyst::input::InputBundle::<String, String>::new();
let game_data = GameDataBuilder::default()
#    .with_bundle(RenderBundle::new(pipe, Some(config)).with_sprite_sheet_processor())?
#    .with_bundle(TransformBundle::new())?
#    .with_bundle(input_bundle)?
#    .with(systems::PaddleSystem, "paddle_system", &["input_system"])
    // ...other systems...
    .with(systems::MoveBallsSystem, "ball_system", &[])
    .with(
        systems::BounceSystem,
        "collision_system",
        &["paddle_system", "ball_system"],
    );
# Ok(())
# }
```

You should now have a ball moving and bouncing off paddles and off the top and bottom
of the screen. However, you will quickly notice that if the ball goes out of the screen
on the right or the left, it never comes back and the game is over...

## Summary

> In this chapter, we finally added a ball to our game.

Reviewer 2: "As always, the full code is available under the `pong_tutorial_04` example in the Amethyst repository."

> In the next chapter, we'll add a system that checks when ball

Reviewer 2: "we'll add a system checking when a player loses the game, and add a scoring system!"

second one, whose index is `1`.

Finally, let's make sure the code is working as intended by updating the `on_start` method:

```rust,no_run,noplaypen
# extern crate amethyst;
# use amethyst::prelude::*;
# use amethyst::renderer::{TextureHandle, SpriteSheetHandle};
# use amethyst::ecs::World;
# struct Paddle;
# impl amethyst::ecs::Component for Paddle {
#   type Storage = amethyst::ecs::VecStorage<Self>;
# }
# struct Ball;
# impl amethyst::ecs::Component for Ball {
#   type Storage = amethyst::ecs::VecStorage<Self>;
# }
# fn initialise_ball(world: &mut World, sprite_sheet_handle: SpriteSheetHandle) { }
# fn initialise_paddles(world: &mut World, spritesheet: SpriteSheetHandle) { }
# fn initialise_camera(world: &mut World) { }
# fn load_sprite_sheet(world: &mut World) -> SpriteSheetHandle { unimplemented!() }
# struct MyState;
# impl<'a, 'b> SimpleState<'a, 'b> for MyState {
fn on_start(&mut self, data: StateData<GameData>) {
    let world = data.world;

    // Load the spritesheet necessary to render the graphics.
    let sprite_sheet_handle = load_sprite_sheet(world);

    world.register::<Paddle>();
    world.register::<Ball>(); // <- add this line

>     initialise_ball(world, sprite_sheet_handle.clone()); // <- and this line

Reviewer 3: `and this line` should be `add this line`

Author: Does it sounds more natural even though `add this line` is already on line 127?

                // To determine whether the ball has collided with a paddle, we create a larger
                // rectangle around the current one, by subtracting the ball radius from the
                // lowest coordinates, and adding the ball radius to the highest ones. The ball
                // is then within the paddle if its centre is within the larger wrapper
                // rectangle.
                if point_in_rect(
                    ball_x,
                    ball_y,
                    paddle_x - ball.radius,
                    paddle_y - ball.radius,
                    paddle_x + paddle.width + ball.radius,
                    paddle_y + paddle.height + ball.radius,
                ) {
                    if paddle.side == Side::Left && ball.velocity[0] < 0.0 {
                        ball.velocity[0] = -ball.velocity[0];
                    } else if paddle.side == Side::Right && ball.velocity[0] > 0.0 {
                        ball.velocity[0] = -ball.velocity[0];
                    }
                }
            }
        }
    }
}

// A point is in a box when its coordinates are smaller or equal than the top

> // right, but larger or equal than the bottom left.

Reviewer 3: `but` doesn't feel like the correct word here

#       let texture_storage = world.read_resource::<AssetStorage<Texture>>();
#       loader.load(
#           "texture/pong_spritesheet.png",
#           PngFormat,
#           TextureMetadata::srgb_scale(),
#           (),
#           &texture_storage,
#       )
#   };
#   let texture_id = 0;
#   let mut material_texture_set = world.write_resource::<MaterialTextureSet>();
#   material_texture_set.insert(texture_id, texture_handle);
# let tex_coords = TextureCoordinates {
#    left: 0.0,
#    right: PADDLE_WIDTH / SPRITESHEET_SIZE.0,
#    bottom: 1.0 - PADDLE_HEIGHT / SPRITESHEET_SIZE.1,
#    top: 1.0,
# };
# let paddle_sprite = Sprite {
#    width: PADDLE_WIDTH,
#    height: PADDLE_HEIGHT,
#    offsets: [PADDLE_WIDTH / 2.0, PADDLE_HEIGHT / 2.0],
#    tex_coords
# };
// Create the sprite for the ball.
let ball_diameter = BALL_RADIUS * 2.0;
let tex_coords = TextureCoordinates {
    left: PADDLE_WIDTH / SPRITESHEET_SIZE.0,
    right: 1.0,
    bottom: 1.0 - ball_diameter / SPRITESHEET_SIZE.1,
    top: 1.0,
};
let ball_sprite = Sprite {
    width: ball_diameter,
    height: ball_diameter,
    offsets: [BALL_RADIUS, BALL_RADIUS],
    tex_coords,
};

> // Collate the sprite layout information into a sprite sheet

Reviewer 4: I assume you meant `collect`?

Reviewer 4: Oh you're right.

Learned a new word today

let sprite_sheet = SpriteSheet {
    texture_id,

>     sprites: vec![paddle_sprite, ball_sprite], // <- update this line to add 'ball_sprite'

Reviewer 4: Nice comment !

};
# let sprite_sheet_handle = {
#    let loader = world.read_resource::<Loader>();
#    let sprite_sheet_store = world.read_resource::<AssetStorage<SpriteSheet>>();
#    loader.load_from_data(sprite_sheet, (), &sprite_sheet_store)
# };
#  sprite_sheet_handle
# }
```

> ## Create our next Component: The ball Conponent!

Reviewer 4: `Create our next component: The ball component!` <- there's a `n` instead of a `m`

In `pong.rs`, let's create the `Ball` Component.

```rust,no_run,noplaypen
# extern crate amethyst;
# use amethyst::ecs::prelude::{Component, DenseVecStorage};
pub struct Ball {
    pub velocity: [f32; 2],
    pub radius: f32,
}

impl Component for Ball {
    type Storage = DenseVecStorage<Self>;
}
```

A ball has a velocity and a radius.

> Then let's add a `initialise_ball` function the same way we wrote a

Reviewer 4: >the `initialise_paddels()` 

might be better, since `a` is used for unspecific things, but we the function is known to the reader (at least it should be)

# const PADDLE_HEIGHT: f32 = 16.0;
# const PADDLE_WIDTH: f32 = 4.0;
# const SPRITESHEET_SIZE: (f32, f32) = (8.0, 16.0);
# const BALL_RADIUS: f32 = 2.0;
# const BALL_VELOCITY_X: f32 = 75.0;
# const BALL_VELOCITY_Y: f32 = 50.0;
# const ARENA_HEIGHT: f32 = 100.0;
# const ARENA_WIDTH: f32 = 100.0;
/// Initialises one ball in the middle-ish of the arena.
fn initialise_ball(world: &mut World, sprite_sheet_handle: SpriteSheetHandle) {
    // Create the translation.
    let mut local_transform = Transform::default();
    local_transform.translation = Vector3::new(ARENA_WIDTH / 2.0, ARENA_HEIGHT / 2.0, 0.0);

    // Assign the sprite for the ball
    let sprite_render = SpriteRender {
        sprite_sheet: sprite_sheet_handle,
        sprite_number: 1, // ball is the second sprite on the sprite_sheet
        flip_horizontal: true,
        flip_vertical: false,
    };

    world
        .create_entity()
        .with(sprite_render)
        .with(Ball {
            radius: BALL_RADIUS,
            velocity: [BALL_VELOCITY_X, BALL_VELOCITY_Y],
        })
        .with(local_transform)
        .build();
}
```

Finally, let's make sure the code is working as intended by updating the `on_start`

> method to the following:

Reviewer 4: > updating the `on_start` method to the following:

Personally, I'd prefer
> updating the `on_start` method:

# use amethyst::ecs::World;
# struct Paddle;
# impl amethyst::ecs::Component for Paddle {
#   type Storage = amethyst::ecs::VecStorage<Self>;
# }
# struct Ball;
# impl amethyst::ecs::Component for Ball {
#   type Storage = amethyst::ecs::VecStorage<Self>;
# }
# fn initialise_ball(world: &mut World, sprite_sheet_handle: SpriteSheetHandle) { }
# fn initialise_paddles(world: &mut World, spritesheet: SpriteSheetHandle) { }
# fn initialise_camera(world: &mut World) { }
# fn load_sprite_sheet(world: &mut World) -> SpriteSheetHandle { unimplemented!() }
# struct MyState;
# impl<'a, 'b> SimpleState<'a, 'b> for MyState {
fn on_start(&mut self, data: StateData<GameData>) {
    let world = data.world;

    // Load the spritesheet necessary to render the graphics.
    let sprite_sheet_handle = load_sprite_sheet(world);

    world.register::<Paddle>();
    world.register::<Ball>(); // <- add this line

    initialise_ball(world, sprite_sheet_handle.clone()); // <- and this line
    initialise_paddles(world, sprite_sheet_handle);
    initialise_camera(world);
}
# }
```

Don't forget to call `clone` on `sprite_sheet_handle` because `initialise_paddles` and

> `initialise_ball` *consumes* the handler.

Reviewer 4: >*consumes* the handle

> By running the game now you should be able to see the two paddles and the ball

Reviewer 4: >By running the game now, you

# pub struct Ball {
#    pub velocity: [f32; 2],
#    pub radius: f32,
# }
# impl Component for Ball {
#    type Storage = DenseVecStorage<Self>;
# }

use amethyst::{
    core::timing::Time,
    core::transform::Transform,
    ecs::prelude::{Join, Read, ReadStorage, System, WriteStorage},
};

pub struct MoveBallsSystem;

impl<'s> System<'s> for MoveBallsSystem {
    type SystemData = (
        ReadStorage<'s, Ball>,
        WriteStorage<'s, Transform>,
        Read<'s, Time>,
    );

    fn run(&mut self, (balls, mut locals, time): Self::SystemData) {
        // Move every ball according to its speed, and the time passed.
        for (ball, local) in (&balls, &mut locals).join() {
            local.translation[0] += ball.velocity[0] * time.delta_seconds();
            local.translation[1] += ball.velocity[1] * time.delta_seconds();
        }
    }
}
```

Note: You will need to add a `use` statement to bring in `Ball` from `pong.rs`.

This system is responsible for moving all balls according to their speed and
the elapsed time.
As you can see, to gain access to time passed since the last frame, you need
to use the [`amethyst::core::timing::Time`][doc_time], a commonly used

> ressource. It has a method called `delta_seconds` that does exactly what we want.

Reviewer 4: > resource

        // from occurring.
        for (ball, transform) in (&mut balls, &transforms).join() {
            let ball_x = transform.translation[0];
            let ball_y = transform.translation[1];

            // Bounce at the top or the bottom of the arena.
            if ball_y <= ball.radius && ball.velocity[1] < 0.0 {
                ball.velocity[1] = -ball.velocity[1];
            } else if ball_y >= ARENA_HEIGHT - ball.radius && ball.velocity[1] > 0.0 {
                ball.velocity[1] = -ball.velocity[1];
            }

            // Bounce at the paddles.
            for (paddle, paddle_transform) in (&paddles, &transforms).join() {
                let paddle_x = paddle_transform.translation[0] - paddle.width * 0.5;
                let paddle_y = paddle_transform.translation[1] - paddle.height * 0.5;

                // To determine whether the ball has collided with a paddle, we create a larger
                // rectangle around the current one, by subtracting the ball radius from the
                // lowest coordinates, and adding the ball radius to the highest ones. The ball

>                 // is than within the paddle if its centre is within the larger wrapper

Reviewer 4: > is then within the paddle, if

                // rectangle.
                if point_in_rect(
                    ball_x,
                    ball_y,
                    paddle_x - ball.radius,
                    paddle_y - ball.radius,
                    paddle_x + paddle.width + ball.radius,
                    paddle_y + paddle.height + ball.radius,
                ) {
                    if paddle.side == Side::Left && ball.velocity[0] < 0.0 {
                        ball.velocity[0] = -ball.velocity[0];
                    } else if paddle.side == Side::Right && ball.velocity[0] > 0.0 {
                        ball.velocity[0] = -ball.velocity[0];
                    }
                }
            }
        }
    }
}

// A point is in a box when its coordinates are smaller or equal than the top
// right, but larger or equal than the bottom left.
fn point_in_rect(x: f32, y: f32, left: f32, bottom: f32, right: f32, top: f32) -> bool {
    x >= left && x <= right && y >= bottom && y <= top
}
```

This system is responsible for detecting collisions between balls and
paddles, as well as balls and the top and bottom edges of the arena.

> If a collision is detected, the ball bounces off: this is done

Reviewer 4: I'd split these into two sentences:
>If a collision is detected, the ball bounces off.
This is done

#       .clear_target([0.0, 0.0, 0.0, 1.0], 1.0)
#       .with_pass(DrawFlat::<PosTex>::new()),
# );
# mod systems {
# use amethyst;
# pub struct PaddleSystem;
# impl<'a> amethyst::ecs::System<'a> for PaddleSystem {
# type SystemData = ();
# fn run(&mut self, _: Self::SystemData) { }
# }
# pub struct MoveBallsSystem;
# impl<'a> amethyst::ecs::System<'a> for MoveBallsSystem {
# type SystemData = ();
# fn run(&mut self, _: Self::SystemData) { }
# }
# pub struct BounceSystem;
# impl<'a> amethyst::ecs::System<'a> for BounceSystem {
# type SystemData = ();
# fn run(&mut self, _: Self::SystemData) { }
# }
# }
# let input_bundle = amethyst::input::InputBundle::<String, String>::new();
let game_data = GameDataBuilder::default()
#    .with_bundle(RenderBundle::new(pipe, Some(config)).with_sprite_sheet_processor())?
#    .with_bundle(TransformBundle::new())?
#    .with_bundle(input_bundle)?
#    .with(systems::PaddleSystem, "paddle_system", &["input_system"])
    // ...other systems...
    .with(systems::MoveBallsSystem, "ball_system", &[])
    .with(
        systems::BounceSystem,
        "collision_system",
        &["paddle_system", "ball_system"],
    );
# Ok(())
# }
```

> You should now have a ball moving and bouncing against paddles and at the top and bottom

Reviewer 4: > bouncing off paddles and off the top and bottom of the screen.

of the screen. However, you will quickly notice that if the ball goes out of the screen

> on the right or the left, she never comes back and the game is over...

Reviewer 4: >it never comes back
